common/djangoapps/student/tests/test_models.py

```diff
@@ -244,6 +256,169 @@ def test_enrollments_not_deleted(self):
+                UserCelebration.perform_streak_updates(self.user, self.course_key)
+                self.assertEqual(self.user.celebration.longest_ever_streak, i)
+
+    # lint-amnesty, pylint: disable=line-too-long
+    def test_celebrate_only_once_with_multiple_calls_on_the_same_day(self):
+        """
+        Sample run for a 3 day streak and 1 day break. See last column for explanation.
+        +---------+---------------------+--------------------+-------------------------+------------------+-----------------------------------------------+
+        | today   | first_day_of_streak | last_day_of_streak | last day of celebration | should_celebrate | Note                                          |
+        +---------+---------------------+--------------------+-------------------------+------------------+-----------------------------------------------+
+        | 2/4/21  | 2/4/21              | 2/4/21             | None                    | FALSE            | Day 1 of Streak                               |
+        | 2/4/21  | 2/4/21              | 2/4/21             | None                    | FALSE            | Day 1 of Streak                               |
+        | 2/5/21  | 2/4/21              | 2/5/21             | None                    | FALSE            | Day 2 of Streak                               |
+        | 2/5/21  | 2/4/21              | 2/5/21             | None                    | FALSE            | Day 2 of Streak                               |
+        | 2/6/21  | 2/4/21              | 2/6/21             | 2/6/21                  | TRUE             | Completed 3 Day Streak so we should celebrate |
+        | 2/6/21  | 2/4/21              | 2/6/21             | 2/6/21                  | FALSE            | Already celebrated this streak.               |
+        +---------+---------------------+--------------------+-------------------------+------------------+-----------------------------------------------+
+        """
+        now = datetime.datetime.now(UTC)
+        for i in range(1, STREAK_LENGTH_TO_CELEBRATE + 1):
+            with mock.patch.object(UserCelebration, '_get_now') as get_now_mock:
+                get_now_mock.return_value = now + datetime.timedelta(days=i)
+                should_celebrate = UserCelebration.perform_streak_updates(self.user, self.course_key)
+                self.assertEqual(should_celebrate, i == STREAK_LENGTH_TO_CELEBRATE)
+                should_celebrate = UserCelebration.perform_streak_updates(self.user, self.course_key)
+                self.assertEqual(should_celebrate, False)
+
+    # lint-amnesty, pylint: disable=line-too-long
+    def test_celebration_with_utc(self):
+        """
+        Check that the _get_now method uses the user's timezone
+        """
+        set_user_preference(self.user, 'time_zone', 'Asia/Tokyo')
+        with mock.patch('crum.get_current_request', return_value=self.request):
```

Comment:
You can also call `crum.set_current_request` to use less mocked code. I believe other tests have done that too.

Reply:
> is the rationale that less mocked code is better?

I think my functional-test bias was just showing. Fuctional vs unit is a sliding scale, but I lean more towards functional. I think of it like "functional tests finds bugs, unit tests tell you where they are" -- and if you can only have one, I'd rather find bugs.

That said, if you are going to mock out your code boundaries, I'd think you'd want to mock `user_timezone_locale_prefs` rather than `crum`. But maybe that's too far towards the unit side of the scale for your own tastes.

>  i guess it wouldn't affect other tests in this case since we're mocking the request on setUp, though it could if we were mocking the request on setUpClass for example, in which case that could have unintended side effects?

Look at other test that use `set_current_request` and hopefully they also do an `addCleanup` call to clear the request. But maybe that's more work than just mocking it, indeed.

Reply:
sure i guess using the cleanup works fine, changed to using set_current_request and addCleanup